File: obel/obeltests/aabb.m
//
//  aabb.m
//  obeltests
//
//  Created by Jack Mousseau on 9/22/21.
//

#import <obel/aabb.h>
#import <XCTest/XCTest.h>

@interface ObelAABBTests : XCTestCase

@end

@implementation ObelAABBTests

- (void)testAABBLayout {
    obel_aabb_t a = { 1.f, 2.f, 3.f, 4.f, 5.f, 6.f };
    XCTAssertEqual(a.min.x, 1.f);
    XCTAssertEqual(a.min.y, 2.f);
    XCTAssertEqual(a.min.z, 3.f);
    XCTAssertEqual(a.max.x, 4.f);
    XCTAssertEqual(a.max.y, 5.f);
    XCTAssertEqual(a.max.z, 6.f);
}

- (void)testAABBVolume {
    obel_aabb_t a = { 1.f, 3.f, 2.f, 4.f, 5.f, 6.f };
    XCTAssertEqual(obel_aabb_volume(a), 24.f);
}

- (void)testAABBIntersectionOverUnion {
    obel_aabb_t a = { 1.f, 2.f, 3.f, 11.f, 22.f, 33.f };
    obel_aabb_t b = { 6.f, 7.f, 8.f, 16.f, 27.f, 38.f };
    XCTAssertEqual(obel_aabb_iou(a, b), 1875.f / (12000.f - 1875.f));
}

- (void)testAABBUnion {
    obel_aabb_t a = { 1.f, 2.f, 3.f, 11.f, 22.f, 33.f };
    obel_aabb_t b = { 6.f, 7.f, 8.f, 16.f, 27.f, 38.f };
    obel_aabb_t aabb = obel_aabb_union(a, b);
    XCTAssertEqual(aabb.min.x,  1.f);
    XCTAssertEqual(aabb.min.y,  2.f);
    XCTAssertEqual(aabb.min.z,  3.f);
    XCTAssertEqual(aabb.max.x, 16.f);
    XCTAssertEqual(aabb.max.y, 27.f);
    XCTAssertEqual(aabb.max.z, 38.f);
}

@end

File: obel/aabb.c
//
//  aabb.c
//  obel
//
//  Created by Jack Mousseau on 9/22/21.
//

#include <obel/aabb.h>

// MARK: - Metrics

OBEL_AABB_FUNC float obel_aabb_volume(obel_aabb_t a) {
    return obel_vector_reduce_mul(obel_vector_sub(a.max, a.min));
}

OBEL_AABB_FUNC float obel_aabb_iou(obel_aabb_t a, obel_aabb_t b) {
    float ab_volume = obel_aabb_volume(obel_aabb_intersection(a, b));
    return ab_volume / (obel_aabb_volume(a) + obel_aabb_volume(b) - ab_volume);
}

// MARK: - Intersection

OBEL_AABB_FUNC obel_aabb_t obel_aabb_intersection(obel_aabb_t a, obel_aabb_t b) {
    obel_vector3_t min = obel_vector_max(a.min, b.min);
    obel_vector3_t max = obel_vector_min(a.max, b.max);
    return (obel_aabb_t){ min, max };
}

OBEL_AABB_FUNC obel_aabb_t obel_aabb_union(obel_aabb_t a, obel_aabb_t b) {
    obel_vector3_t min = obel_vector_min(a.min, b.min);
    obel_vector3_t max = obel_vector_max(a.max, b.max);
    return (obel_aabb_t){ min, max };
}

OBEL_AABB_FUNC obel_aabb_t obel_aabb_union(obel_aabb_t a, obel_vector3_t b) {
    obel_vector3_t min = obel_vector_min(a.min, b);
    obel_vector3_t max = obel_vector_max(a.max, b);
    return (obel_aabb_t){ min, max };
}
